src/storage-client/src/sink.rs

```diff
@@ -250,62 +258,63 @@ pub async fn build_kafka(
         KafkaConsistencyConfig::Progress { topic } => {
-            meta.topics().iter().find(|t| t.name() == topic)
-        }
-    };
-
-    // If the consistency topic exists, check to see if it contains this sink's
-    // data.
-    let latest_ts = if let Some(progress_topic) = progress_topic {
-        let progress_client: BaseConsumer<_> = connection
-            .connection
-            .create_with_context(
-                connection_cx,
-                MzClientContext::default(),
-                &btreemap! {
-                    "group.id" => SinkGroupId::new(sink_id),
-                    "isolation.level" => "read_committed".into(),
-                    "enable.auto.commit" => "false".into(),
-                    "auto.offset.reset" => "earliest".into(),
-                    "enable.partition.eof" => "true".into(),
-                },
+            ensure_kafka_topic(
+                &client,
+                topic,
+                1,
+                connection.replication_factor,
+                KafkaSinkConnectionRetention::default(),
             )
-            .await?;
-
-        let progress_client = Arc::new(progress_client);
-        let latest_ts = determine_latest_progress_record(
-            format!("build_kafka_{}", sink_id),
-            progress_topic.name().to_string(),
-            ProgressKey::new(sink_id),
-            Arc::clone(&progress_client),
-        )
-        .await
-        .check_ssh_status(progress_client.client().context())?;
-
-        // If we have progress data, we should have the topic listed in the
-        // broker's metadata. If we don't, error.
-        if latest_ts.is_some() && !meta.topics().iter().any(|t| t.name() == connection.topic) {
-            Err(anyhow::anyhow!(
-                "sink progress data exists, but sink data topic is missing"
-            ))?
-        }
+            .await
+            .check_ssh_status(client.inner().context())
+            .add_context("error registering kafka consistency topic for sink")?;
+
+            let progress_client: BaseConsumer<_> = connection
+                .connection
+                .create_with_context(
+                    connection_cx,
+                    MzClientContext::default(),
+                    &btreemap! {
+                        "group.id" => SinkGroupId::new(sink_id),
+                        "isolation.level" => "read_uncommitted".into(),
+                        "enable.auto.commit" => "false".into(),
+                        "auto.offset.reset" => "earliest".into(),
+                        "enable.partition.eof" => "true".into(),
+                    },
+                )
+                .await?;
+
+            let progress_client = Arc::new(progress_client);
+            let latest_ts = determine_latest_progress_record(
+                format!("build_kafka_{}", sink_id),
+                topic.to_string(),
+                ProgressKey::new(sink_id),
+                Arc::clone(&progress_client),
```

Comment:
Wait, it's slightly more subtle than just passing in the `read_uncommitted` consumer, right? As written I think there's a possibility for incorrectness here, which is that if the prior sink has written out a progress message but _not_ committed it, we'll restart, see that progress message and resume from the _next_ timestamp... despite the fact that sink has never actually committed a transaction for that timestamp!

I think we need to push the `progress_client` creation into `determine_latest_progress_record` and construct _two_ progress clients: one in `read_uncommitted` mode to fetch the high water mark, plus one in `read_committed` mode to actually read the progress messages out—potentially blocking if there are outstanding transactions until those transactions commit or abort.

Reply:
Oh, that is totally correct and what Petros had talked about. I missed this in the flurry of commits and branches that this work became. Thank you for the catch here––I'll get this in on Monday.

Reply:
Pushed this fix––not very ergonomic so very open to style suggestions.